src/AstGen.zig

```diff
@@ -2138,6 +2144,34 @@ fn genDefers(
     }
 }
 
+fn checkUsed(
+    gz: *GenZir,
+    outer_scope: *Scope,
+    inner_scope: *Scope,
+) InnerError!void {
+    const astgen = gz.astgen;
+
+    var scope = inner_scope;
+    while (scope != outer_scope) {
+        switch (scope.tag) {
+            .gen_zir => scope = scope.cast(GenZir).?.parent,
+            .local_val => {
+                const s = scope.cast(Scope.LocalVal).?;
+                if (!s.used) return astgen.failTok(s.token_src, "unused local constant", .{});
+                scope = s.parent;
+            },
+            .local_ptr => {
+                const s = scope.cast(Scope.LocalPtr).?;
+                if (!s.used) return astgen.failTok(s.token_src, "unused local variable", .{});
```

Comment:
Yes, I think I will have a followup pr that will make used a `enum { fn_param, constant, variable, loop_capture, loop_enumeration, capture, used }` so that we can emit **really** nice errors. I didn't want to introduce more complexity in this pr, but if you want, I can.